syntax="proto2";

package npctransport_proto;

message Configuration {
  message XYZ {
  // XYZ coordinates for a particle etc.
    required double x=1;
    required double y=2;
    required double z=3;
  };
  message IntRange {
    // a range from lower to upper with specified number of steps
    required int32 lower=1;
    optional int32 upper=2;
    optional int32 steps=3;
  };
  message FloatRange {
    // a range from lower to upper with specified number of steps
    // evenly-spaced on a logarithmic space with specified base.
    // if base == 1, standard uniform spacing is used
    required double lower=1;
    optional double upper=2;
    optional int32 steps=3;
    optional double base=4;
  };
  message FGConfiguration {
    required IntRange number_of_beads=1;
    required IntRange number=2;
    required FloatRange radius=3;
    required IntRange interactions=4;
    required FloatRange rest_length_factor=5;
    required FloatRange d_factor=6;
    required FloatRange interaction_k_factor=7;
    required FloatRange interaction_range_factor=8;
    repeated XYZ anchor_coordinates=9; // a fixed coordinate for the first bead in the chain
  };
  message FloatConfiguration {
    required IntRange number=1;
    required FloatRange radius=2;
    required IntRange interactions=3;
    required FloatRange d_factor=4;
    required FloatRange interaction_k_factor=7;
    required FloatRange interaction_range_factor=8;
  };
  message Interaction {
    required string type0=1;
    required string type1=2;
    required IntRange is_on=3;
    optional FloatRange interaction_k=4;
    optional FloatRange interaction_range=5;
  };

  required FloatRange interaction_k=3;
  required FloatRange interaction_range=4;
  required FloatRange backbone_k=5;
  required FloatRange time_step_factor=6;
  required FloatRange box_side=7;
  required IntRange box_is_on=8;
  required FloatRange slab_thickness=9;
  required FloatRange tunnel_radius=10;
  required IntRange slab_is_on=11;
  required FloatRange slack=12;
  required int32 number_of_trials=13;
  required double maximal_number_of_frames=14;
  repeated FGConfiguration fgs=15;
  repeated FloatConfiguration floaters=16;
  repeated Interaction interactions=17;
  required double dump_interval_ns=18; // 0 = every frame
  required FloatRange nonspecific_range=19;
  required FloatRange nonspecific_k=20;
  required FloatRange angular_D_factor=21;
  required double statistics_interval_ns=22; // 0 = every frame
  required FloatRange excluded_volume_k=23;
  required double simulation_time_ns=24;
  // percent of time in actual running (as opposed to equlibration):
  required FloatRange statistics_fraction=25;
  optional double maximum_number_of_minutes=26;
}

// if you add any parameters you must update automatic_parameters.cpp
// accordingly
message Assignment {
  message XYZ {
  // XYZ coordinates for a particle etc.
    required double x=1;
    required double y=2;
    required double z=3;
  };
  message IntAssignment {
    required int32 value=1;
    optional int32 index=2;
  };
  message FloatAssignment {
    required double value=1;
    optional int32 index=2;
  };
  message FGAssignment {
    required IntAssignment number_of_beads=1;
    required IntAssignment number=2;
    required FloatAssignment radius=3;
    required IntAssignment interactions=4;
    required FloatAssignment rest_length_factor=5;
    required FloatAssignment d_factor=6;
    required FloatAssignment interaction_k_factor=7;
    required FloatAssignment interaction_range_factor=8;
    repeated XYZ anchor_coordinates=9; // a fixed coordinate for the first bead in the chain
  };
  message FloaterAssignment {
    required IntAssignment number=1;
    required FloatAssignment radius=2;
    required IntAssignment interactions=3;
    required FloatAssignment d_factor=4;
    required FloatAssignment interaction_k_factor=5;
    required FloatAssignment interaction_range_factor=7;
  };
  message InteractionAssignment {
    required string type0=1;
    required string type1=2;
    required IntAssignment is_on=3;
    optional FloatAssignment interaction_k=4;
    optional FloatAssignment interaction_range=5;
  };
  required int32 work_unit=1;
  required FloatAssignment interaction_k=3;
  required FloatAssignment interaction_range=4;
  required FloatAssignment backbone_k=5;
  required FloatAssignment time_step_factor=6;
  // length of the side of a box
  required FloatAssignment box_side=7;
  required IntAssignment box_is_on=8;
  required FloatAssignment slab_thickness=9;
  required FloatAssignment tunnel_radius=10;
  required IntAssignment slab_is_on=11;
  required FloatAssignment slack=12;
  required int32 number_of_trials=13;
  required double maximal_number_of_frames=14;
  repeated FGAssignment fgs=15;
  repeated FloaterAssignment floaters=16;
  repeated InteractionAssignment interactions=17;
  required double dump_interval_ns=18; // 0 = every frame
  required FloatAssignment nonspecific_range=19;
  required FloatAssignment nonspecific_k=20;
  required FloatAssignment angular_D_factor=21;
  required double statistics_interval_ns=22; // 0 = every frame
  required FloatAssignment excluded_volume_k=23;
  required int32 number_of_frames=24; // = simulation_time_ns * 1e+6 / time_step
  required double time_step=25; // in femtoseconds (IMP default)
  required double range=26;
  required double simulation_time_ns=27;
  required double statistics_interval_frames=28; // = statistics_interval_ns * 1e+6 / time_step
  required double dump_interval_frames=29; // = dump_interval_ns * 1e+6 / time_step
  required FloatAssignment statistics_fraction=30;
  optional double maximum_number_of_minutes=31;
  optional uint64 random_seed = 32; // random seed used to initialize this run
}

message Statistics {
  message FGStats {
    optional double volume=1 [default=0];
    optional double length=2 [default=0];
    optional double radius_of_gyration=3 [default=0];
    optional double particle_diffusion_coefficient=4 [default=0];
    optional double particle_correlation_time=5 [default=0];
    optional double chain_diffusion_coefficient=6 [default=0];
    optional double chain_correlation_time=7 [default=0];
    optional double local_diffusion_coefficient=8 [default=0];
  }
  message FloatStats {
    optional double interactions=1 [default=0];
    optional double interacting=2 [default=0];
    optional double interaction_partner_chains=3 [default=0];
    optional double interaction_partner_beads=4 [default=0];
    optional double diffusion_coefficient=5 [default=0];
    optional double correlation_time=6 [default=0];
    optional double avg_n_transports=7 [default=0]; // avg number of times a particle of this type has crossed the channel
    repeated double transport_time_points_ns=8; // records the time points of transports in simulation time (TODO: assuming single trial?)
  }
  message InteractionStats {
    required string type0=1;
    required string type1=2;
    optional double avg_contacts_per_particle0=3;
    optional double avg_contacts_per_particle1=4;
    optional double avg_pct_bound_particles0=5;
    optional double avg_pct_bound_particles1=6;
  }
  repeated FGStats fgs=1;
  repeated FloatStats floaters=2;
  repeated InteractionStats interactions=6;
  optional double energy_per_particle=3 [default=0];
  optional int32 number_of_frames=4 [default=0]; // number of frames for which statistics were gathered
  optional double bd_simulation_time_ns=8 [default=0]; // number of ns in which BD simulation was running
  optional double seconds_per_iteration=5 [default=0]; // overall time in seconds for this statistics iteration
  optional int32 interrupted=7 [default=0];
}

message Conformation {
  message Coordinates {
    required double x=1;
    required double y=2;
    required double z=3;
  };
  message Sites {
     required string name=1;
     repeated Coordinates coordinates=2;
  };
  repeated Sites sites=1;
  message Particle {
    required double x=1;
    required double y=2;
    required double z=3;
    required double r=4;
    required double i=5;
    required double j=6;
    required double k=7;
  };
  repeated Particle particle=2;
};

message Output {
  required Assignment assignment=1;
  required Statistics statistics=2;
  optional Conformation conformation=3;
  optional bytes rmf_conformation=4;
}
